pages/api/events/class/index.ts

```diff
@@ -57,28 +87,81 @@ const classEventHandler: NextApiHandler = async (req: NextApiRequest, res: NextA
 
           const dateEnd = DateTime.fromJSDate(dateWithoutTime)
             .set({
               hour: endTime.hour,
               minute: endTime.minute,
               second: endTime.second,
             })
-            .setZone(newEvent.timeZone);
-
-          intervals.push(Interval.fromDateTimes(dateStart, dateEnd));
+            .setZone(newEvent.timeZone, { keepLocalTime: true });
+
+          const interval = Interval.fromDateTimes(dateStart, dateEnd);
+          intervals.push(interval);
+
+          const weekDay = dateStart.weekday;
+          if (!seenWeekDays.has(weekDay)) {
+            seenWeekDays.add(weekDay);
+            const weekDayStr = indexToWeekdays[weekDay] as Weekdays;
+            if (weekDayStr != "sun") {
+              intervalsByWeekDay[weekDayStr] = interval;
+            }
+          }
         }
 
-        // verify that each teacher is available during class times
+        // verify scheduling for each teacher
         try {
           for (const teacher of teachers) {
+            // verify that teacher doesn't have another event during this class
             await validateTimes(teacher, intervals);
+
+            // verify that teacher has availability set during this class time
+            if (newEvent.checkAvailabilities) {
+              const availabilities = await getAvailabilityById(teacher.id);
+              let weekday: keyof WeekdayIntervals;
+              for (weekday in intervalsByWeekDay) {
+                const interval = intervalsByWeekDay[weekday];
+                let isAvailable = false;
+                if (availabilities && interval) {
+                  const availabilityTimes = availabilities[weekday];
+                  if (availabilityTimes) {
+                    for (const time of availabilityTimes) {
+                      const availableStartDateTime = timeDateZoneToDateTime(
+                        time[0],
+                        interval.start,
+                        availabilities.timeZone
+                      );
+                      const availableEndDateTime = timeDateZoneToDateTime(
+                        time[1],
+                        interval.start,
+                        availabilities.timeZone
+                      );
+                      const availabilityInterval = Interval.fromDateTimes(
+                        availableStartDateTime,
+                        availableEndDateTime
+                      );
+                      if (availabilityInterval.engulfs(interval)) {
+                        isAvailable = true;
+                        break;
+                      }
+                    }
+                  }
+                }
+                if (interval && !isAvailable) {
+                  throw new TeacherAvailabilityError(
+                    `Teacher ${teacher.firstName} ${teacher.lastName} is not available for class ${newEvent.name}`
+                  );
+                }
+              }
+            }
           }
```

Comment:
lets move this to another function since this is is pretty messy. It should be like the `validateTimes()` function (especially since this will probably be used over and over agin for new types of events)

Reply:
Also, some commenting throughout :)